hudi-common/src/main/java/org/apache/hudi/common/model/HoodieKey.java

```diff
@@ -31,13 +35,16 @@
 
   private String recordKey;
   private String partitionPath;
+  private List<Object> indexKey;
 
-  public HoodieKey() {
+  public HoodieKey(String recordKey, String partitionPath) {
+    this(recordKey, partitionPath, Collections.emptyList());
   }
 
-  public HoodieKey(String recordKey, String partitionPath) {
+  public HoodieKey(String recordKey, String partitionPath, List<Object> indexKey) {
```

Comment:
Index key is used to tag location and cluster the insert/upsert data. To avoid repeated serialization and deserialization, can we store the index key and save it in `HoodieRecord`?